query-algebrizer/src/clauses/mod.rs

```diff
@@ -825,7 +826,139 @@ impl ConjoiningClauses {
     /// where `?y` must simultaneously be a ref-typed attribute and a boolean-typed attribute. This
     /// function deduces that and calls `self.mark_known_empty`. #293.
     pub fn expand_type_tags(&mut self) {
-        // TODO.
+        for (v, tag) in self.extracted_types.iter() {
+            let cols = self.column_bindings.get(v).unwrap().clone();
+            for col in cols {
+                if col.0 != tag.0 { 
+                    let value = QueryValue::Column(QualifiedAlias(col.0.clone(), Column::Fixed(DatomsColumn::ValueTypeTag)));
+                    let alias = tag.clone();
+                    self.wheres.0.push(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(alias, value)));
+                }
+            }
+        }
+
+        let mut computed_types: BTreeMap<Variable, TypedValue> = BTreeMap::default();
+
+        for computed_table in self.computed_tables.iter() {
+            match computed_table {
+                &ComputedTable::NamedValues { ref names, ref values } => {
+                    for (i, item) in names.iter().enumerate() {
+                        computed_types.insert(item.clone(), values[i].clone());
+                    }
+                },
+                _ => (),
+            }
+        }
+
+        let mut empty_reason: Option<EmptyBecause> = None;
+        let mut wheres_to_add = vec![];
+        for (v, value_type_set) in self.known_types.iter() {
+            if self.column_bindings.contains_key(v) {
+                let cols = self.column_bindings.get(v).unwrap().clone();
+                for col in cols {
+                    match col.1 {
+                        Column::Fixed(DatomsColumn::Value) => { 
+                            // attempt to figure out exactly which type is used
+                            // TODO: But only if the attribute is a variable or not present and therefore of unknown type
+                            let mut value_type: Option<ValueType> = None;
+                            for constraint in self.wheres.0.iter() {
```

Comment:
There are three variables in that query.

- `?y` is the attribute, so it's an entity (`ValueType::Ref`).
- `?x` is the 'e' of the first pattern, so it's also a ref.
- `?z` is the 'e' of the second pattern, so it's also a ref.

Processing the rest of the pattern, we see that `?y` is used again in the second pattern in the same way — still ref. `?x` is used again as the 'v' of the second pattern, which doesn't change anything.

We conclude that all of `?x`, `?y`, and `?z` must be refs, and we don't need to project a type from the SQL tables.

The thing we need to conclude at this point, for this issue, is that `datoms01`, constrained by `datoms01.v = datoms00.e AND datoms01.a = datoms00.a`, can yield bindings for `?x` that are not refs, but share a value-space representation with refs: `1` could match `true`, `1970-01-01…00.001`, etc.

As such we need to additionally constrain by type.